### DISCUSSION TITLE
How can I clear banned users except for ignored ones?

### DISCUSSION THREAD

**Discussion by ghost** [nicotine-plus/nicotine-plus#2424] [Q&A] [answered]
Instead of doing this
```python
for list_name in ("banlist", "ignorelist", "ipblocklist", "ipignorelist"):
    self.config.sections["server"][list_name].clear()
```
I want to unban only the users that aren't ignored. So loop through the `banlist` checking the names in the `ignorelist` and loop the `ipblocklist` checking the ips in the `ipignorelist` and clear only the users that aren't ignored. How can I do this?

It's probably something like this
```python
# Get the names and ips from the banlist and ipblocklist
banned_names = self.config.sections["server"]["banlist"].keys()
banned_ips = self.config.sections["server"]["ipblocklist"].keys()

# Get the names and ips from the ignorelist and ipignorelist
ignored_names = self.config.sections["server"]["ignorelist"].keys()
ignored_ips = self.config.sections["server"]["ipignorelist"].keys()

# Loop through the banned names and clear them if they are not in the ignorelist
for name in banned_names:
    if name not in ignored_names:
        self.config.sections["server"]["banlist"].clear(name)

# Loop through the banned IPs and clear them if they are not in the ipignorelist
for ip in banned_ips:
    if ip not in ignored_ips:
        self.config.sections["server"]["ipblocklist"].clear(ip)
```

**Comment by slook** [comment, marked as answer]
[clear()](https://docs.python.org/3/library/stdtypes.html#dict.clear) is a function which removes all items from a list or dictionary, so it doesn't take any arguments.

Instead, you could use `remove()` to delete a specific key such as:
- `self.config.sections["server"]["banlist"].remove(name)`

Alternatively, you can call:
- `self.core.networkfilter.unban_user(name)`

To unban an IP (even if you don't have the IP address):
-  `self.core.networkfilter.unban_user_ip(name[, address])`

**Reply by ghost** [reply]
The del option doesn't work. I get a `TypeError: list indices must be integers or slices, not str`. I'm going to try the remove.

**Reply by slook** [reply]
It's probably better to call the unban functions in the networkfilter module, because there might be some caching whereby modifying the config directly won't have the proper effect.

**Reply by ghost** [reply]
Oh right I've tried the remove option and it seems to work but I'll change it for the unban function in the networkfilter module. Thanks.

**Comment by Inso-m-niaC** [comment]
I knocked a method up that should do the job
```python
    def purge_bans(self):
    
        ignorelist = self.config.sections["server"]["ignorelist"]
        ipignorelist = self.config.sections["server"]["ipignorelist"]
        
        ignored_users = ignorelist + [x for x in ipignorelist.values() if x and x not in ignorelist]
        ignored_ips = ipignorelist.keys()

        for ip_address, user in self.config.sections["server"]["ipblocklist"].items():
            if user in ignored_users or ip_address in ignored_ips:
                continue
            try:
                self.core.network_filter.unban_user_ip(user, ip_address)
            except AttributeError:
                self.core.network_filter.unblock_user_ip(user, ip_address)
            
            self.log("Unbanning %s: %s" % (ip_address, user))

        for user in self.config.sections["server"]["banlist"]:
            if user in ignored_users:
                continue
            self.core.network_filter.unban_user(user)
            self.log("Unbanning %s" % user)
```
